allennlp/training/trainer.py

```diff
@@ -526,11 +526,11 @@ def _should_stop_early(self, metric_history: List[float]) -> bool:
             # Pylint can't figure out that in this branch `self._patience` is an int.
             # pylint: disable=invalid-unary-operand-type
 
-            # Is the best score in the past N epochs worse than the best score overall?
+            # Is the best score in the past N epochs worse than or equal the best score overall?
             if self._validation_metric_decreases:
-                return min(metric_history[-self._patience:]) > min(metric_history)
+                return min(metric_history[-self._patience:]) >= min(metric_history[:-self._patience])
```

Comment:
Actually, it's not necessary to compute the min/max of ``history[:-patience]``: if the best value is not in ``history[-patience:]``, then it must be the element ``history[-patience - 1]``, otherwise we should have stopped at the end of the previous epoch.

Of course, this is a negligible optimization and I guess you can leave the code as it is.